bundles/com.e1c.dt.check.form/src/com/e1c/dt/check/form/InvalidItemIdCheck.java

```diff
@@ -0,0 +1,420 @@
+
+    /**
+     * Creates new instance.
+     * @param invalidItemIdService Service that will be used to delegate checks to. Must not be {@code null}.
+     */
+    @Inject
+    public InvalidItemIdCheck(IInvalidItemIdService invalidItemIdService)
+    {
+        Objects.requireNonNull(invalidItemIdService, "invalidItemIdService"); //$NON-NLS-1$
+        this.invalidItemIdservice = invalidItemIdService;
+    }
+
+    @Override
+    public String getCheckId()
+    {
+        return IInvalidItemIdService.CHECK_ID;
+    }
+
+    @Override
+    protected void configureCheck(CheckConfigurer configurationBuilder)
+    {
+        configurationBuilder.extension(new SkipBaseFormExtension())
+            .extension(new TopObjectFilterExtension())
+            .extension(new AdditionalRevalidationRules())
+            .title(Messages.InvalidItemIdCheck_title)
+            .description(Messages.InvalidItemIdCheck_description)
+            .complexity(CheckComplexity.NORMAL)
+            .severity(IssueSeverity.MAJOR)
+            .issueType(IssueType.ERROR)
+            .topObject(FormPackage.Literals.FORM)
+            .checkTop();
+    }
+
+    @Override
+    protected void check(Object object, ResultAcceptor resultAcceptor, ICheckParameters parameters,
+        IProgressMonitor progressMonitor)
+    {
+        if (!(object instanceof Form))
+        {
+            // We will get all kinds of child content here due to call
+            // to ICheckDefinition.addCheckedModelObjects()
+            // in InvalidItemIdCheck.AdditionalRevalidationRules.configureContextCollector().
+            // We're not insterested in this child content here and need to ignore all those calls.
+            return;
+        }
+        if (progressMonitor.isCanceled())
+        {
+            CorePlugin.trace(IInvalidItemIdService.DEBUG_OPTION, "Check: Check has been cancelled"); //$NON-NLS-1$
+            return;
+        }
+        Form form = (Form)object;
+        invalidItemIdservice.validate(form)
+            .entrySet()
+            .stream()
+            .forEach(itemAndMesage -> resultAcceptor.addIssue(itemAndMesage.getValue(), itemAndMesage.getKey(),
+                FormPackage.Literals.FORM_ITEM__ID));
+    }
+
+    /**
+     * Finds parent form of the given object.
+     *
+     * @param bmObject Object for which to find parent form. Must not be {@code null}.
+     * @return Parent form of the given object or an empty value
+     * if there is no top object or top object is not a form. Never {@code null}.
+     *
+     * @see IBmObject#bmGetTopObject()
+     */
+    private static Optional<Form> findFormOf(IBmObject bmObject)
+    {
+        IBmObject topObject = bmObject.bmGetTopObject();
+        if (!(topObject instanceof Form))
+        {
+            CorePlugin.trace(IInvalidItemIdService.DEBUG_OPTION,
+                "Check: Unable to determine parent form: object={0}, topObject={1}", //$NON-NLS-1$
+                bmObject, bmObject.bmGetTopObject());
```

Comment:
С точки зрения API, будет либо NPE, либо CCE:

```java
    /**
     * Gets the corresponding top-level object. If the object is not attached to BM an exception is thrown.
     *
     * @return the corresponding top-level object, can be {@code null} in case if {@link IBmObject} is global and
     * the object was removed out of storage.
     */
    IBmObject bmGetTopObject();
```

Reply:
Код трассировки был удален.